### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by mr-darker** [mr-darker/MDB-C#14]

**Comment by mr-darker** [comment]
> I can see how ESP family would be useful

It's actually the price of the ESP32 WROOM module, which is usable in a custom design, which is why i picked the ESP32.  At the time, I was only making a peripheral so didn't want much computing power.

>I'm interested in any open vending software/hardware effort. We can start by creating Github organization and some communication channel

Awesome!  Well my code is organized in a specific way.  I uploaded an info-graphic that describes how the files are organized.  In general, it's set up where you can follow along with the manual as you work on the code.  Or it's supposed to be if it isn't.  This is still a work in progress so keep that in mind.

>What kind of help you need?

The most useful help would be simple C advice/expertise, from someone who understands the MDB protocol, as well as what my goals/intentions are in the code as a whole.  The rest is sorta just manual labor/data-entry transcribing the manual over to the appropriate C files in the same format used by all the other files, which I imagine isn't a job others wanna do, which is okay because I'm probably the best fit for that and it's straightforward work that I can do myself.  General improvements/additions would be nice too.  Say if you adapted the code to work on your hardware, adding the file for it to work on a RaspberryPi would be a nice addition.  Really just any sorta help anyone wants to contribute, helps, even just providing  an outside view with opinions on how things should be done differently.

>Are you developing a vending machine?

I actually do hope to develop a VMC eventually.  After this code is finished I intend to do the same with ALL other vending communication standards.  Then get started on a VMC, which is a business venture, but will be open source either the day I fail, or the day I start selling the VMC I make, whichever day comes first.  That's been the plan anyways.  
I haven't done any work on a VMC yet though.  There is a VMC file within my MDB code, but each MDB device is a "Blank" device.  The cashless payment device doesn't come with some bitcoin payment service or credit card reader, unless you code it to have payment options it has none.  The VMC doesn't come with any functions like motor controls or anything outside MDB communications.  Each device is like a "blank check" and doesn't offer functionality that isn't described in the MDB manual.  If you don't program functionality into each device file, all it'll do is just ACK polls all day.
The idea is that people take this software as a "base" for their specific application, and have abstract commands to use that require little to no knowledge of the MDB standard.  Basically I don't want anyone to have to read the MDB manual anymore.

**Comment by temoto** [comment]
> The most useful help would be simple C advice/expertise, from someone who understands the MDB protocol

Anything concrete?

Right away I'd point to GCC `__builtin_parity()` which may compile to something faster than your.

>  Say if you adapted the code to work on your hardware, adding the file for it to work on a RaspberryPi would be a nice addition.

TL;DR: it's not quite an "adding the file" task, but is rather big decision on how to work with UART on the VMC part. Or I didn't understand your perspective well enough.

I was motivated to avoid additional hardware and make 9bit serial directly on Raspberry. Tried these:
- Linux tty with parity dance -- parity setting switching was too slow. Maybe it's possible to fix with kernel scheduler settings.
- Linux 4.9+ tty with CMSPAR -- you have to issue only one syscall to switch from bit9-high to bit9-low to send command from VMC to peripherals. After receiving response from peripheral, VMC must respond with ACK/NAK within 5ms. In our case it was 4.2-4.5ms - too close to deadline. Again, maybe possible to fix with kernel scheduler config. I've tried some process priority and sysctl settings without any change.
- Linux module using almost same code as user-space -- it works, but saves only some hundred microseconds. Which is great news - Linux syscalls are cheap. But still too close to MDB deadline.
- Linux module talking directly to UART hardware -- I'm not competent in kernel enough, failed to find a way to grab UART hardware instead of special kernel code.
- software GPIO bit-banging using https://github.com/joan2937/pigpio works nicely. I've extracted pigpio to separate process written in Rust (to have some safety guarantees). MDB response latency is about 2.3-2.8ms which is sub-optimal by far, but well within safety margin. Again, it's likely possible to make this latency close to zero with kernel scheduler settings.

Out of these options, your code is probably best fit for low-level Linux driver. And I think such module would be generally very useful and appreciated. But we need help of a Linux kernel hardware guy. Quick googling showed that serial/tty code is special unlike other device drivers and intercepting its job is not easy.

**Comment by temoto** [comment]
> even just providing an outside view with opinions on how things should be done differently.

In our VMC project I strictly decided against C code in the high-level business logic part. Because I suck at C and make stupid mistakes like off-by-1, buffer overflows. Unfortunately, you can't program ATMega in Rust yet, I would happily pick that option.

General advice is to separate low-level hardware stuff and high-level business logic stuff and pick the best tool for each.

Just recently I finally understood high-level design structure of our VMC:
- scenario execution engine, provides API like "execute sequence (or tree) of commands"
- device drivers, provide new commands to execution engine
- configuration and scenarios to make a particular vending machine -- I'm thinking on declarative config file like graphviz dot and GUI editor; because it's the part where human understanding and easy changes are most important

**Comment by mr-darker** [comment]
>Anything concrete?

Well I've been getting help off a C forum which works for immediate issues.  I saw you using a bunch of different languages so assumed you know a lot more than me about C, which still is probably true technically.  I can use google and forums to find answers to questions, I was looking more for answers to questions I didn't have yet, like pointing out something obviously/terribly wrong with the way I chose to do something.  Occasionally I do get stuck on an issue where I can't seem to find an answer though so maybe I'll ask you if I'm not having luck.

**In response to the 9 bit options you listed:**
If you didn't know, Real-Time-Linux is a thing.  I've never used it, but from what I read, a RTOS adds easy control over the exact time things/processes happen in the code.  Maybe it's unnecessary though, I don't know much about Linux kernel.  The best possible solution I can think of for taking care of MDB's communication formatting, is programming a small FPGA to be a "MDB Bridge Chip".  This has been my long term plan (still using Linux, but with an FPGA to connect to MDB stuff), but the more I think about it, the more I wanna do it right meow, and the more it makes sense to do it right meow.   I think I'll focus on developing some VHDL code for such an adapter, make & sell the adapter to whoever needs/wants one (probably $1-5 each, however low I can get it) then remove all of the code that it takes care of, from the repo and have that be the standard way to use this code (still saving the old code incase there's a rebel who doesn't wanna use the adapter).  Maybe I'll do that tomorrow, maybe not.  Tomorrow I'm cooking chicken fried rice, and I cook like 20 pounds at a time, so it usually takes me all day.


>Your code is probably best fit for low-level Linux driver.

I plan to port it over to Linux when it's finished, as a module.  That'll be my learning project for learning to use Linux kernel stuff.  But that's why I'm sticking to C.  Also because I'm a hardware guy and like how it's as close to the hardware as you can get before you get into Assembly or Machine Language which takes a bit longer to do simple stuff, and isn't portable across different hardware.


> In our VMC project I strictly decided against C code in the high-level business logic part.

Me too, well the plan was to use python or something on the other side, that and/or Linux script stuff.  I'm not exactly sure how fancy features/software would prefer to interact with my MDB code, but I do know that abstraction and simplicity is desired, and that's what the goal is of this MDB project, abstract, simple, modular/configurable, and universal to any hardware.  I also suck at C, but I'm sticking to it because of the reasons I mentioned earlier, also it produces the most efficient code (next to assembly, which isn't cross-platform portable), and the longer I stick to it, the better I'll get at it, and then hopefully if all my life plans fail, I can at least land a programming job some day lol.

**Comment by mr-darker** [comment]
might just be another late night idea too (making the adapter), we'll see tomorrow when I think about it again

**Comment by temoto** [comment]
One company made MDB converter with PIC microcontroller (cheaper and less powerful than ATMega we used). https://blog.abrantix.com/webshop/product/mdb-converter-3/

If they didn't put a ridiculous price on it, we'd buy it, save so much time and be happy.

I don't know your way, for me learning is the opposite of getting product done.

**Comment by mr-darker** [comment]
MDB stuff is extremely overpriced, partially due to the low amount of customers that want it, and partially due to it not being super simple to make an MDB device (which is what this project aims to fix).  That device you sent a link to probably could be sold for $10 and still bring in a profit.  

I put 3 images of a device I was making [on my website](Darker.Tech).  I scrapped that project due to family issues, my family's vending business was the reason I was working on it.  I got that down to $11 each (not including the housing) in low quantities.  So, I'm confident I could make a MDB adapter for about $5, and sell them between $5-10 each, depending on quantity people buy (shipping not included).  I don't really want profits for this, but don't want to end up donating too much of my time to assembling circuit boards and fulfilling orders.

For me, learning-by-doing is one of the best ways to spend your time working on a (personal) project.  If everything fails, you are left with new skills/experience you can apply to your next projects.  Even if you lose, you still win, and you never waste your time this way.  You just gotta think of the knowledge/experience as your guaranteed payoff, and the project's success as a bonus you may or may not get.

**Comment by temoto** [comment]
Looks nice. What interface is on another end? You scrapped that project and now continue? Or it was production ready?

**Comment by mr-darker** [comment]
It's a peripheral, so MDB on both ends.  Was going to be a Cashless Payment Device, connecting the machine to wifi/bluetooth.  It was [production ready](https://www.youtube.com/watch?time_continue=9&v=1mK7gysYJpE), but still a prototype technically.  I just wanted about 100 units for my father's vending busienss, so I could start automating the inventorying, and someone else I knew was working on a bitcoin payment service startup company, he's the one that wanted to sell it.  He and I parted ways though when I scrapped it, I gave him the files and code and all that, and I don't think he had success with his business.  The bitcoin market dropped/crashed some time after that when china or whoever made it illegal to trade in their country.

I had the VMC idea a long time ago, but decided to make the smaller device so I could get simpler features sooner rather than later, for my father's business.  I scrapped it and am back to working on a bigger project, making a "universal" VMC.  Moreso, preparing to work on it.  It's a lot cheaper to work on software, so I'm getting all the software ready before working on hardware.  Starting with this MDB code.
